### PR TITLE
Convert to typescript

### PR DESCRIPTION
**Pull request by Page-**
The two conversions combined work out to an ~91% performance improvement for the optimisation + compilation step in the tests, with the improvement being split roughly evenly between the two steps.

The output between the ometajs and typescript versions should be equivalent (all existing tests continue to pass in the same way, and resin-api tests also pass) and the code is fairly close to a 1-1 mapping.  The main changes are that the optimisation step now handles some cases where we don't require the first element to be a string denoting the node type and inserts the correct type - handling those cases lets the compilation step be simpler and paves the way to removing support for those cases altogether

### CODE REVIEW

File: AbstractSQLOptimiser.ts [hunk 1]
```diff
@@ -0,0 +1,1217 @@
...
+	try {
+		return JSONValue(args);
+	} catch (e) {
+		CheckErr(e);
+	}
+	try {
+		return DurationValue(args);
+	} catch (e) {
+		CheckErr(e);
+	}
```
Comment (zvin) [inline AbstractSQLOptimiser.ts]: This could be written with a `for valueType of [TextValue, NumericValue, ...] {`

File: AbstractSQLRules2SQL.ts [hunk 2]
```diff
@@ -0,0 +1,1165 @@
...
+
+export type Binding = [string, any] | ['Bind', number | string | any[]];
+export interface SqlResult {
+	query: string;
+	bindings: Binding[];
+}
+
+type MatchFn = (args: AbstractSqlQuery, indent: string) => string;
+
+const comparisons = {
```
Comment (zvin) [inline AbstractSQLRules2SQL.ts]: this is already declared in AbstractSQLOptimiser.ts

File: AbstractSQLRules2SQL.ts [hunk 3]
```diff
@@ -0,0 +1,1165 @@
...
+			}
+		}
+		return '$' + fieldOrderings.push(bind);
+	} else {
+		fieldOrderings.push(bind);
+		return '?';
+	}
+};
+
+let fieldOrderings: Binding[] = [];
```
Comment (thgreasi) [inline AbstractSQLRules2SQL.ts]: Let's move this before `AddBind` so that it's not used before it's declaration

File: AbstractSQLRules2SQL.ts [hunk 4]
```diff
@@ -0,0 +1,1165 @@
...
+let engine: Engines = Engines.postgres;
```
Comment (thgreasi) [inline AbstractSQLRules2SQL.ts]: Let's move this one as well a bit higher, near the top of the file.

File: AbstractSQLRules2SQL.ts [hunk 5]
```diff
@@ -0,0 +1,1165 @@
...
+					break;
+				case 'Limit':
+					if (limit !== '') {
+						throw new SyntaxError(
+							`'SelectQuery' can only accept one '${type}'`,
+						);
+					}
+					limit = indent + typeRules[type](rest, indent);
+					break;
+				case 'Offset':
```
Comment (thgreasi) [inline AbstractSQLRules2SQL.ts]: Seems like Where, GroupBy, OrderBy, Limit & Offset could be using an abstract implementation

File: AbstractSQLRules2SQL.ts [hunk 6]
```diff
@@ -0,0 +1,1165 @@
...
+				: '';
+
+		return (
+			'SELECT ' + select + from + where + groupBy + orderBy + limit + offset
+		);
+	},
+	Select: (args, indent) => {
+		checkArgs('Select', args, 1);
+		args = getAbstractSqlQuery(args, 0);
+		if (args.length === 0) {
```
Comment (thgreasi) [inline AbstractSQLRules2SQL.ts]: Should we use `_.isEmpty()` to also protect from null/undefined?

File: AbstractSQLOptimiser.ts [hunk 7]
```diff
@@ -0,0 +1,1217 @@
...
+	name: string,
+	matchers: MatchFn[],
+	rewriteFn: RewriteFn,
+): MatchFn => args => {
+	checkArgs(name, args, matchers.length);
+	return rewriteFn(
+		args.map((arg, index) => {
+			// if (!_.isArray(arg)) {
+			// 	throw new SyntaxError(`Expected AbstractSqlQuery array but got ${typeof arg}`)
+			// }
```
[the remark this answers is not in the document]
Comment (Page-) [reply AbstractSQLOptimiser.ts]: Nope, removed

File: AbstractSQLRules2SQL.ts [hunk 8]
```diff
@@ -0,0 +1,1165 @@
...
+const CheckErr = (e: Error) => {
+	if (e instanceof SyntaxError || e instanceof TypeError) {
+		throw e;
+	}
+};
+
+const escapeField = (field: string | AbstractSqlQuery) =>
+	field === '*' ? '*' : `"${field}"`;
+
+const AnyValue: MatchFn = (args, indent) => {
```
[the remark this answers is not in the document]
Comment (Page-) [reply AbstractSQLRules2SQL.ts]: Sort of, they should be equivalent (they operate on the same input) but one redirects the the optimisation functions and the other to the compilation functions, ie one returns an `AbstractSqlQuery` to go through more optimisation passes as needed or to then be passed to the compiler, whilst the other returns a SQL string

File: AbstractSQLRules2SQL.ts (hunk 6 above)
Comment (Page-) [reply AbstractSQLRules2SQL.ts]: `getAbstractSqlQuery` guarantees we're working with an array, so there's no need to guard against null/undefined

File: AbstractSQLCompiler.ts [hunk 9]
```diff
@@ -1,3 +1,9 @@
+export enum Engines {
```
[the remark this answers is not in the document]
Comment (Page-) [reply AbstractSQLCompiler.ts]: I checked and for this case and the object is still generated/exported (which is good) but it does change
```ts
let engine: Engines = Engines.postgres;
```
from
```js
var engine = AbstractSQLCompiler_1.Engines.postgres;
```
to
```js
var engine = "postgres";
```

File: AbstractSQLRules2SQL.ts [hunk 10]
```diff
@@ -0,0 +1,1165 @@
...
+	GreaterThanOrEqual: ' >= ',
+	LessThan: ' < ',
+	LessThanOrEqual: ' <= ',
+	NotEquals: ' != ',
+	Like: ' LIKE ',
+};
+
+const NestedIndent = (indent: string): string => indent + '\t';
+
+const CheckErr = (e: Error) => {
```
[the remark this answers is not in the document]
Comment (Page-) [reply AbstractSQLRules2SQL.ts]: For `comparisons` we only need its type in the optimiser, so I left it in here and just imported in the optimiser (the import isn't in the generated js), and for `CheckErr` with other changes it's now only used in one place so I just inlined it

**Comment by Page-** [comment]
@zvin the typescript files aren't generated, but they are used for transforming the abstract sql and them compiling it to sql - so the sql the typescript version generates is the same as the ometajs version and so is compatible/only a minor change, but all of the typescript is handwritten and comments are definitely useful

**Comment by Page-** [comment]
The main reason for the change was the (immensely) improved performance, but it also gives us better error messages (ie actual error messages with some information instead of just `match failed`) and a better ability for others to work with the code (since learning ometa takes a lot of time and doesn't have any real docs).
In terms of effort to work with the ometa vs typescript versions, I think for me the ometajs does win out slightly but only because I've had so much experience with it and that the abstract sql tree was largely designed to be easy for ometa to work with. And I think the accessibility for others is more important than that

File: AbstractSQLCompiler.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (thgreasi) [reply AbstractSQLCompiler.ts]: > the object is still generated/exported

Oh, that's probably b/c we use `"preserveConstEnums": true`:
https://github.com/balena-io-modules/abstract-sql-compiler/blob/59b932725f51ea68f3634beabd03f93d4190e279/tsconfig.json#L7

[the remark this answers is not in the document]
Comment (Page-) [reply AbstractSQLCompiler.ts]: Ah, sweet, that makes life easier for non-ts interop :)

**Comment by Page-** [comment]
I pushed some additional changes to avoid throwing errors as part of normal matching (throwing errors is expensive) and also deduplicated some code between compiler/optimiser (the `FooValue` checks)

**Comment by Page-** [comment]
Not easily @dfunckt, it would in theory be possible to do something where you have the same traversal code for both and then different actions for the two, but passing around the indentation level makes that more awkward, as does the fact that the optimising step accepts more forms of AST than the compiler (largely due to backwards compatibility) which it then rewrites to forms the compiler accepts.  Definitely something that can be done but I think it would be a fairly significant piece of work atm

**Pull request merged**